# Sebelum menuliskan code nya, harus membuat formula matrix nya terlebih dahulu, supaya lebih mudah membuat program/fungsinya

Kali ini permasalahan nya menggunakan matrix orde 4 sesuai dengan persoalan Fisika nomor 25

In [1]:
# misal: 
# x1 = a
# x2 = T1
# x3 = T2
# x4 = T3

## Dari penyederhanaan persamaannya, didapatkan matrix yaitu:
# 10x1 + x2 + 0 + 0 = [[98.2*(sin45 - 0.25*cos45)
# 4x1 - x2 + x3 + 0 = [39.28*(sin45 - 0.3*cos45)]
# 5x1 + 0 - x3 + x4 = [49.1*(sin45 - 0.2*cos45)]
# 6x1 + 0 + 0 - x4 = [-58.92]
#Ruas kiri untuk matriX A dan ruas kanan untuk matrix B

In [3]:
#Input modul yang dibutuhkan
#Modul numpy digunakan untuk membuat matrixnya
#modul math digunakan untuk menghitung sin dan cos
import numpy as np
import math 

#Agar lebih mudah, sin45 dan cos45 dihitung dan dikonversi, lalu di deklarasikan terlebih dahulu 
sin45 = math.sin(math.radians(45))
cos45 = math.cos(math.radians(45))

#Input matrix A
a = [[10.0, 1.0, 0.0, 0.0],
     [4.0, -1.0, 1.0, 0.0],
     [5.0, 0.0, -1.0, 1.0],
     [6.0, 0.0, 0.0, -1.0]]

#Input matrix B
b = [[98.2*(sin45 - 0.25*cos45)],
    [39.28*(sin45 - 0.3*cos45)],
    [49.1*(sin45 - 0.2*cos45)],
    [-58.92]]

#Menjadikan a dan b berupa matrix
a = np.array(a)
b = np.array(b)

#Membuat Fungsi Gauss Eliminan
def gaussElimin(a,b):
    n = len(b)
    # Phase Elimination
    for k in range(0,n-1):
        for i in range(k+1,n):
            if a[i,k] != 0.0:
                lam = a [i,k]/a[k,k]
                a[i,k+1:n] = a[i,k+1:n] - lam*a[k,k+1:n]
                b[i] = b[i] - lam*b[k]
    # Back substitution
    for k in range(n-1,-1,-1):
        b[k] = (b[k] - np.dot(a[k,k+1:n],b[k+1:n]))/a[k,k]
    return b

#Mencetak hasilnya
print("Matrix A = ")
print(a)
print("Matrix B = ")
print(b)
print("") #String pemisah
print("Hasil perhitungan menggunakan metode Gauss Elimination:")
print(gaussElimin(a,b)) #Pemanggilan fungsi gaussEliminan

Matrix A = 
[[10.  1.  0.  0.]
 [ 4. -1.  1.  0.]
 [ 5.  0. -1.  1.]
 [ 6.  0.  0. -1.]]
Matrix B = 
[[ 52.07841443]
 [ 19.44260806]
 [ 27.77515437]
 [-58.92      ]]

Hasil perhitungan menggunakan metode Gauss Elimination:
[[ 1.61504707]
 [35.92794369]
 [48.91036345]
 [68.61028245]]


In [4]:
#Agar lebih mudah, sin45 dan cos45 dihitung dan dikonversi, lalu di deklarasikan terlebih dahulu 
sin45 = math.sin(math.radians(45))
cos45 = math.cos(math.radians(45))

#Input matrix A
a = [[10.0, 1.0, 0.0, 0.0],
     [4.0, -1.0, 1.0, 0.0],
     [5.0, 0.0, -1.0, 1.0],
     [6.0, 0.0, 0.0, -1.0]]

#Input matrix B
b = [[98.2*(sin45 - 0.25*cos45)],
    [39.28*(sin45 - 0.3*cos45)],
    [49.1*(sin45 - 0.2*cos45)],
    [-58.92]]

#Menjadikan a dan b berupa matrix
a = np.array(a)
b = np.array(b)

#Membuat fungsi untuk LU Decomposition
def LUdecomp(a):
    n = len(a)
    for k in range(0,n-1):
        for i in range(k+1,n):
            if a[i,k] != 0.0:
                lam = a [i,k]/a[k,k]
                a[i,k+1:n] = a[i,k+1:n] - lam*a[k,k+1:n]
                a[i,k] = lam
    return a

#Membuat fungsi untuk LUsolve
def LUsolve(a,b):
    n = len(a)
    for k in range(1,n):
        b[k] = b[k] - np.dot(a[k,0:k],b[0:k])
    b[n-1] = b[n-1]/a[n-1,n-1]
    for k in range(n-2,-1,-1):
        b[k] = (b[k] - np.dot(a[k,k+1:n],b[k+1:n]))/a[k,k]
    return b

#Mencetak hasil nya
print("Hasil untuk LU decomposition:")
print(LUdecomp(a)) #Pemanggilan fungsi LUdecomp
print("") #String pemisah
print("Hasil untuk LU solve:")
print(LUsolve(a,b))  #Pemanggilan fungsi LUsolve

Hasil untuk LU decomposition:
[[10.          1.          0.          0.        ]
 [ 0.4        -1.4         1.          0.        ]
 [ 0.5         0.35714286 -1.35714286  1.        ]
 [ 0.6         0.42857143  0.31578947 -1.31578947]]

Hasil untuk LU solve:
[[ 1.61504707]
 [35.92794369]
 [48.91036345]
 [68.61028245]]


In [6]:
# Metode Gauss Jordan
#Dengan contoh matrix yg lain
import numpy as np
import math

#Agar lebih mudah, sin45 dan cos45 dihitung dan dikonversi, lalu di deklarasikan terlebih dahulu 
sin45 = math.sin(math.radians(45))
cos45 = math.cos(math.radians(45))

#Input matrix A
a = [[10.0, 1.0, 0.0, 0.0],
     [4.0, -1.0, 1.0, 0.0],
     [5.0, 0.0, -1.0, 1.0],
     [6.0, 0.0,  0.0,-1.0]]

#Input matrix B
b = [[98.2*(sin45 - 0.25*cos45)],
     [39.28*(sin45 - 0.3*cos45)],
     [49.1*(sin45 - 0.2*cos45)],
     [-58.92]]

#Menjadikan a dan b berupa matrix
a = np.array(a)
b = np.array(b)

#Membuat fungsi untuk Gauss-Jordan
def jordan(a,b):
    n = len(b)
    # Proses Eliminasi segitiga bawah
    for i in range(1,n):
        for j in range(0,n):
            if(i == j or i < j) : continue
            #print(i,j)
            if a[i,j] != 0.0:
                lam = a[i,j]/a[j,j]
                a[i,j:n] = a[i,j:n] - lam*a[j,j:n]
                b[i] = b[i] - lam*b[j]
            if (a[i,j] > -1.0e-6 and a[i,j] < 1.0e-6 ):
                a[i,j] = 0

    # Proses Eliminasi segitiga atas
    for i in reversed(range(0,n-1)):
        for j in reversed(range(i+1,n)):
            if a[i,j] != 0.0:
                lam = a[i,j]/a[j,j]
                a[i,j:n] = a[i,j:n] - lam*a[j,j:n]
                b[i] = b[i] - lam*b[j]
            if (a[i,j] > -1.0e-6 and a[i,j] < 1.0e-6 ):
                a[i,j] = 0      

    # Proses Diagonalisasi
    for i in range(0,n):
        for j in range(0,n):
            if(a[i,i] != 1):
                b[i] /= a[i,i]
                a[i,i] /= a[i,i]
    return b
                
np.set_printoptions(precision=2)
print("Matrix A = ")
print(a)
print("Matrix B = ")
print(b)
print(" ") #String pemisah
print("Hasil perhitungan menggunakan metode Gauss-Jordan:")

print(jordan(a,b))

Matrix A = 
[[10.  1.  0.  0.]
 [ 4. -1.  1.  0.]
 [ 5.  0. -1.  1.]
 [ 6.  0.  0. -1.]]
Matrix B = 
[[ 52.08]
 [ 19.44]
 [ 27.78]
 [-58.92]]
 
Hasil perhitungan menggunakan metode Gauss-Jordan:
[[ 1.62]
 [35.93]
 [48.91]
 [68.61]]


Alhamdullilah persoalan Fisika nomor 25 dari buku Numerical Methods in Engineering with Python sudah terselesaikan menggunakan metode; Gauss-elimination, LU Decomposition-Solve dan Gauss-Jordan. Walaupun dengan cara yang berbeda, namun tetap memberikan hasil yang sama, ini membuktikan bahwa metode yang digunakan sudah benar dan valid.
Didapatkan hasil nya yaitu a = 1.62 ms^-2 , T1 = 35.93 N , T2 = 48.91 N , T3 = 68.61 N